[Timers] Add msec rules timer + loop timer

TD-er opened the pull request:
This adds the following:
- timerSet_ms
- loopTimerSet
- loopTimerSet_ms

The ones with _ms postfix are using msec resolution, while the other ones use seconds as resolution.

The `loopTimerSet` function can create a timer which will re-schedule itself again at exactly the set interval (based on its initial intended execution time)
For example, if running the traditional way of re-scheduling the timer inside the rules block, you may add some offset on every execution interval.
This is typically 20 msec, but may be a lot more on a busy node or when using long rules.

Using the loopTimer, the timers will be scheduled at a constant interval, even though the execution of these events may experience some jitter when some blocking calls are made on the node.

The `timerPause` and `timerResume` commands are also working as their names suggest.

The `Rules#timer` event now has a second value, showing the count.
On single timer events this count is always 0.
On loop timers, this count reflects the number of events since it was (re)created.

A loop timer has a 3rd optional parameter, to set the number of loops until it ends.
When not set, the loop runs forever, until it is stopped or changed.

Fixes: #1614

TD-er committed:
[Rules Timer] Use scheduler for rules timer

This does remove the limit of number of timers.
Also it makes scheduled intervals a lot more accurate.

TD-er committed:
[Timers] Make loop timers with fixed interval and nr of loops

TD-er commented:
With these new commands, you can also create random like timers, or very complex timers.

For example:
```
On System#Boot do    //When the ESP boots, do
  looptimerset_ms,1,2000,10
  looptimerset_ms,2,2500
endon


On Rules#Timer=1 do
  if %eventvalue2% >= 5
    timerSet,1,0
  endif
  //pulse some led on pin 2 shortly
  Pulse,2,0,50
  logentry,%eventvalue2%
endon

On Rules#Timer=2 do
  if %eventvalue2% = 2
    loopTimerSet_ms,2,2500
    timerResume,1
  else
    timerPause,1
  endif
endon
```


If the used loop timers are set to relative prime factors, you can create a randomizer which may take a long time before it starts repeating itself.
Especially when using multiple timers.

Could be used to create a device to simulate flashing of a TV when you're not home.

TD-er commented:
A test build:

- [ESPEasy_ESP32_mega-20200801-23-PR_3204.zip](https://www.dropbox.com/s/ftqm2qlxica2l4x/ESPEasy_ESP32_mega-20200801-23-PR_3204.zip?dl=0)
- [ESPEasy_ESP82xx_mega-20200801-23-PR_3204.zip](https://www.dropbox.com/s/ybqm4v26a19via0/ESPEasy_ESP82xx_mega-20200801-23-PR_3204.zip?dl=0)

TD-er committed:
[Docs] Add documentation for msec timers and loop timers

TD-er committed:
[Timers] Cleanup of timer code.

Wiki591 commented:
So, wih this kind of timer I get an exact  timer for i.e. creating a battery capacity monitoring depending on Ah, which isn't possible up to now by using the normal timer function due to its inaccuracy?

TD-er commented:
With the loop timer you will not see a drift.

For example, if the timer is started at t=1000 msec from boot, with an interval of 1000 msec.
The scheduled actions will be on t=1000, t=2000, etc.

But it still isn't a real-time environment, so it does not mean the timer events will be executed at exactly that moment.
They can be executed at t = 1010, t=2015, t=3001, etc.  
The jitter can be a lot more if there is some blocking code between timer events.
For example at t=2999 a controller starts to send some data which may first timeout on a DNS request and also needs to read data from the flash. This action may take over 1000 msec to run (DNS timeout = 1000 msec) and thus the moments for t=3000 and t=4000 will be missed.
This means the next event will be scheduled for t=5000.

So please have a look at the timing stats page to see what the max. blocking duration is and if it is critical you're not missing a timer event, then use an interval which is larger.
For use cases where it is critical you don't miss such an event, you should set the interval to 10 - 60 seconds as I've seen some calls may really take that long.

For the 'traditional' way of re-scheduling the timer, you will see a drift in the scheduled times.
So the scheduled time will be the last time + jitter + interval, for each call to set the timer.
That's why I created the loop timer variant.

TD-er committed:
[Scheduler] Move global scheduler functions into C++ class

This does make the compilation more predictive and lowers resource usage slightly.  (global functions and objects use slightly more space)

TD-er commented:
As I've added a few changes to create a scheduler object instead of lots of global functions to manage the scheduler, I created a new test build for it.
It shouldn't change functionality compared to the last build, but since it is quite a big change in code, I would like to see it being tested.

- [ESPEasy_ESP32_mega-20200801-28-PR_3204.zip](https://www.dropbox.com/s/yhf0dydu8wsksiy/ESPEasy_ESP32_mega-20200801-28-PR_3204.zip?dl=0)
- [ESPEasy_ESP82xx_mega-20200801-28-PR_3204.zip](https://www.dropbox.com/s/3mrzzbwb44q8m34/ESPEasy_ESP82xx_mega-20200801-28-PR_3204.zip?dl=0)

TD-er committed:
[Timers] Change interval timer defines into enum class

TD-er committed:
[Scheduler] Change plugin pointer type into enum class

Wiki591 commented:
Some time ago I tried to build a battery test environment for getting the remaining useable capacity (Ah) of used, old and reanimated lead acid batteries. I have used the normal Rules Timer but I had to notice, that the timer itself has problems to run correctly. I had to add various correction factors to get a rough estimation of the health and state of the battery. 

Using the LoopTimerSet_ms up to now gives a reasonable information of the battery data. Seems as if at least the LoopTimerSet_ms works as expected. At first I did a calibration run for 1 hour, using my laboratory power supply with 1A of current, result was 100% correct. Then I started a discharging/charging cycle which is running now for about 6 hours, working fine.

Binary used:

```
Build:⋄
20109  - Mega
System Libraries:⋄
ESP82xx Core 5d3af165, NONOS SDK 2.2.2-dev(38a443e), LWIP: 2.1.2 PUYA support
Git Build:⋄

Plugin Count:⋄
46 [Normal]
Build Time:⋄
Aug 10 2020 15:24:01
Binary Filename:⋄
ESP_Easy_mega_20200810_normal_ESP8266_4M1M
Build Platform:⋄
Linux-4.19.104-microsoft-standard-x86_64-with-glibc2.29
Git HEAD:⋄
HEAD

```

Devices used:

![grafik](https://user-images.githubusercontent.com/38845308/89837487-5a39ae00-db69-11ea-9960-21a4f502533f.png)

Rule used:

```
On System#Boot Do
	LoopTimerSet_ms,1,1000
	LoopTimerSet,2,60
EndOn

On Rules#Timer=1 Do
	TaskRun,3
	TaskValueSet 4,1,([BatVolt#Current]/3600)+[AH#AHc]
	TaskValueSet 4,2,[AH#Timer]+(1/3600)
EndOn

On Rules#Timer=2 Do
	Publish domoticz/in,'{"idx":160,"nvalue":0,"svalue":"[AH#AHc]"}'
	Publish domoticz/in,'{"idx":161,"nvalue":0,"svalue":"[AH#Timer]"}'
	Publish domoticz/in,'{"idx":13,"nvalue":0,"svalue":"[BatVolt#Voltage]"}'
	Publish domoticz/in,'{"idx":28,"nvalue":0,"svalue":"[BatVolt#Current]"}'
 	Publish domoticz/in,'{"idx":32,"nvalue":0,"svalue":"[BatVolt#Power]"}'
EndOn
```

Any more test informations needed / wanted?

TD-er commented:
Just to be sure, I think for your use case it is best to use 2 sec interval for timer 1 and thus use 1800 instead of 3600.

Right now I can't think of other things to test, other then try it also with the latest build I linked.
If tomorrow nothing seems to be broken, I can merge it and make a new build for it.

Wiki591 commented:
Thank you for your advice. I know, what you are talking abaout but this project will be ported to my RV to get an estimation of the state of the onboard batteries. These ones are charged by solar panels, so I have to be as close to a realtime data collection as possible (switching of power consuming devices, shadowing by trees, buildings, bridges a.s.o. of the solar panels during driving, clouds, a.s.o.). The device itself will be dedicated to the task of capacity measurement to avoid influences of other blocking tasks and with this testing project I want to get an idea how close I will get to the real remainig capacity.

And up to now it looks pretty good, so I will stay at one second timer.

Btw, I used `ESPEasy_ESP82xx_mega-20200801-28-PR_3204.zip` , I think its your last build?

Wiki591 commented:
Just for documentation purpose:

![grafik](https://user-images.githubusercontent.com/38845308/89841290-dd133680-db72-11ea-81f5-ffd20bcc58e2.png)

![grafik](https://user-images.githubusercontent.com/38845308/89841515-60cd2300-db73-11ea-805b-5f76325ca2e5.png)

![grafik](https://user-images.githubusercontent.com/38845308/89841887-670fcf00-db74-11ea-958b-00406ec9c556.png)

![grafik](https://user-images.githubusercontent.com/38845308/89841714-f799df80-db73-11ea-81f8-1c8a010d81d9.png)

The battery I am charging right now is a worn, twice deep discharged VRLA 140Ah one. There is no constant charging current due to the chemical rebuilding/reanimation inside of the battery - good test for the future purpose.

Wiki591 commented:
Still up and running, very satisfying. Thumbs Up, next step in my RV project will start at the moment the outside temperatures are in an acceptable range.

TD-er commented:
So the only error can be due to non-linear behavior of the measurement.  (shunt heating)
The timing should now be as accurate as possible.

TD-er committed:
[Docs] Update mentioned dates of introduced feature

... to expected build date
